### ISSUE TITLE
Decouple for Multiple Use Cases

### ISSUE THREAD

**Issue by kahboom** [syndesisio/syndesis-rest#471]
## The area for improvement
At the moment, the API is relatively coupled to the backend, not sure if programmatically or just via CI. It would be nice if the API had a feature that allowed for different setups while working with the data it provides.

## Why
The purpose of the API is to provide a layer of abstraction between the UI and the backend. It should be more of a convenience than anything, which helps developers control _what_ data is shared with the UI and backend, _how_ it is shared, its _structure_, etc.

When it isn't, this sometimes leads to slower development in the UI, especially since we are using the staging API to facilitate development (this is for convenience, though, of course, not necessarily the best practice). This may be due to a bug in the API or backend, a feature that has not yet been implemented, maybe some inconsistencies between the UI and API, which are in the process of being resolved, but in the meantime would break staging for demos and make it difficult to work with in the UI.

### Benefits
1. A sort of sandbox area for better testing things out, messing up the data, and trying different things like changes to information structure, without sacrificing what is currently stable and working.
2. Easier to identify bugs, both in the API and the UI, along with areas of inconsistency between the them.
3. Smoother, more predictable demos with more consistent data.
4. Generally, decoupling each area, as long as communication remains consistent across teams, is considered to be a good practice.

### Reasons I think we should not have to do this in the UI
We could (and often have to) stub out dummy data in the UI, but there are several disadvantages to this, and it's a direction I'd rather not continue to go in:

1. We would likely need an ORM for the UI, almost like GraphQL, in order to respect entity relationships and to provide more realistic data, which we should not have to rely on and it can further complicate matters.
2. It makes the UI bulky, slowing down build times for AOT, slowing down the development process, OR initial loading in the browser, sacrificing the user experience. Generally, we want to avoid building a bulky UI under any circumstance, considering we cannot persist data the way the API can, other than in the browser. Adding in multiple sets of dummy JSON data or some sqlite database is a good way to do that for no justifiable reason, especially if we want realistic data and use something like GraphQL.
3. Generating and using dummy data is a PITA, sometimes easier to do manually, which, as I'm sure you can imagine, takes a while to get right. Then we have to adjust by wiring things up to the API once everything is complete, usually really close to (the night before) a demo, where we realize that things are not fully as expected.

### tl;dr;
- A testing and/or development data store, separate from the staging data store.
- A fallback mechanism in either store (haven't thought this one through all the way 😂 ).
- Configuring things in the API for testing and development purposes (with safety precautions).

The longer version. There are a lot of different directions we could go in here, so I'd like to hear from everyone. I propose the following solution, again, is not perfect, but maybe something to work from:

- A separate API-owned data store (last I checked I think were were using PostgreSQL?) for testing purposes, from the one used for staging, as we currently have. I think this is the best solution, because we could use a cleaner setup for demos, and another one for development. I guess this would require a different subdomain and OS cluster? All of this is up for discussion though?
- A fallback mechanism that leverages the data store. For instance, for an endpoint we know will eventually work (or that we temporarily need to get working for development purposes in the UI, or a demo), but is either under maintenance or has yet to be implemented, we could, for instance, return JSON that is automatically provided from a JSON file automatically based on the Swagger spec or, even better, via a SQL data dump in the meantime. How data is stored (e.g. per user), and data cleanup is another discussion.
- Fine-grained configurability from whitelisted users or hosts (any ideas here?) that will allow for configurability in the API. For instance, disabling the creation of GH repos when testing/developing something related to creating creating an integration, if you are not specifically testing for the deployment.

A big consideration to make here is identification of use cases. For instance, sometimes you want a development environment that disables deployments completely, and perhaps the API would just return mocked data that could be manipulated. Other times you might do want to test full user flows, such as with end-to-end testing. Maybe a configuration here would be ideal and set per user session or some configuration parameter, and an example of this is disabling GH repo creation, even disabling/enabling administrative privileges per account when user roles are implemented, and just working with metadata? I think this would only work if we are using a totally separate database.

**Comment by gashcrumb** [comment]
@rhuss wonder, if this issue would be good for discussing coordination between the frontend and backend API usage?  This is at least related to that topic.
